### PR TITLE
Producer topic creation

### PR DESCRIPTION
**Pull request by balexander**
The error we were running into where a topic is created but the first record fails is known. What happens is that the topic is created, but Kafka hasn't done its magic yet with partitions, replication, etc. When the record is then immediately written to the topic there is a `LeaderNotAvailable` error. 

node-rdkafka doesn't have a good way to handle this. My work around is to read the topics from the metadata we get when we first connect to kafka. If the topic doesn't exist then we create it.

### CODE REVIEW

File: index.ts [hunk 1]
```diff
@@ -6,10 +6,16 @@ export type DeliveryReport = {
   key: string | null;
   timestamp: number;
   size: number;
-  value?: Buffer
+  value?: Buffer;
+};
+export type ConnectionData = {
```
Comment (chasdevs) [inline index.ts]: Probably don't need to export this?

File: src/test/consumer/integration/Consumer.test.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import ConsumerBuilder from "../../../consumer/ConsumerBuilder";
+import * as faker from "faker";
+
+describe("Producer integration tests", () => {
```
Comment (chasdevs) [inline src/test/consumer/integration/Consumer.test.ts]: consumer?

File: src/test/producer/unit/Producer.test.ts [hunk 3]
```diff
@@ -95,14 +98,35 @@ describe("Producer Unit Test", () => {
...
-    // await fakeProducer.produce(fakeRecords, topic);
+    // await producer.produce(fakeRecords, topic);
     // td.verify(serializer.encodeMessage(td.matchers.anything(), td.matchers.anything(), td.matchers.anything()), {times:2})
   });
 
+  test("Creates topic if doesn't exist", async () => {
+    const fakeRecord = FakeFactory.fakeRecord();
+    td.when(kafkaClient.connect(td.matchers.anything())).thenCallback(null, { topics: [{ name: "" }] });
+    await producer.produce([fakeRecord]);
+    td.verify(adminClient.createTopic(td.matchers.isA(Object), td.matchers.isA(Function)), { times: 1 });
```
Comment (chasdevs) [inline src/test/producer/unit/Producer.test.ts]: Potentially worth testing that the object contains the fakeRecord topic name

File: src/test/producer/unit/Producer.test.ts [hunk 4]
```diff
@@ -95,14 +98,35 @@ describe("Producer Unit Test", () => {
...
+  });
+
+  test("Does attempt to create topic if it already exists", async () => {
```
Comment (chasdevs) [inline src/test/producer/unit/Producer.test.ts]: does not?

**Comment by chasdevs** [comment]
So the LeaderNotAvailable just isn't handled by node-rdkafka so that's why it was "failing" silently?

**Comment by balexander** [comment]
@chasdevs Right.  First it doesn't have any logic to wait for an automatically created topic.  Second, it treats LeaderNotAvailable, at least when it gets that error in the context of not finding a topic, it just logs what amounts to a warning.

You can see it in the debug logs:
```
{ 
  severity: 7,
  fac: 'METADATA',
  message: '[thrd:main]: ssl://storyblocks-prod-storyblocks-16cb.aivencloud.com:18364/bootstrap:   Topic #0/1: hover-event with 0 partitions: Broker: Leader not available' 
}
{ 
  severity: 7,
  fac: 'METADATA',
  message: '[thrd:main]: ssl://storyblocks-prod-storyblocks-16cb.aivencloud.com:18364/bootstrap: Temporary error in metadata reply for topic hover-event (PartCnt 0): Broker: Leader not available: ignoring' 
}
```

There is some more info in this issue from the confluent python client: https://github.com/dpkp/kafka-python/issues/150

**Pull request closed without merging**